wdl-analysis/src/handlers/goto_definition.rs

```diff
@@ -0,0 +1,604 @@
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use anyhow::Result;
+use anyhow::anyhow;
+use line_index::LineIndex;
+use line_index::WideEncoding;
+use lsp_types::GotoDefinitionResponse;
+use lsp_types::Location;
+use lsp_types::Position;
+use rowan::TextSize;
+use url::Url;
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::Span;
+use wdl_ast::SyntaxKind;
+use wdl_ast::SyntaxNode;
+use wdl_ast::SyntaxToken;
+use wdl_ast::TreeNode;
+use wdl_ast::TreeToken;
+use wdl_ast::v1;
+
+use crate::DiagnosticsConfig;
+use crate::SourcePosition;
+use crate::SourcePositionEncoding;
+use crate::diagnostics;
+use crate::document::Document;
+use crate::document::ScopeRef;
+use crate::graph::DocumentGraph;
+use crate::graph::ParseState;
+use crate::types::v1::EvaluationContext;
+use crate::types::v1::ExprTypeEvaluator;
+
+/// Handler for goto definition requests.
+#[derive(Debug)]
+pub struct GotoDefinitionHandler;
+
+impl Default for GotoDefinitionHandler {
+    fn default() -> Self {
+        Self
+    }
+}
+
+impl GotoDefinitionHandler {
+    /// Creates a new goto definition handler
+    pub fn new() -> Self {
+        Self
+    }
+
+    /// Finds the definition location for an identifier at the given position.
+    ///
+    /// Searches the document and its imports for the definition of the
+    /// identifier at the specified position, returning the location if
+    /// found.
+    ///
+    /// * If a definition is found for the identifier then a
+    ///   [`GotoDefinitionResponse`] containing the URI and range is returned
+    ///   wrapped in [`Some`].
+    ///
+    /// * Else, [`None`] is returned.
+    pub fn goto_definition(
+        &self,
+        graph: &DocumentGraph,
+        document_uri: Url,
+        position: SourcePosition,
+        encoding: SourcePositionEncoding,
+    ) -> Result<Option<GotoDefinitionResponse>> {
+        let index = graph
+            .get_index(&document_uri)
+            .ok_or_else(|| anyhow!("document not found in graph"))?;
+
+        let node = graph.get(index);
+        let (root, lines) = match node.parse_state() {
+            ParseState::Parsed { lines, root, .. } => {
+                (SyntaxNode::new_root(root.clone()), lines.clone())
+            }
+            _ => return Err(anyhow!("document not yet parsed: {}", document_uri)),
+        };
+
+        let Some(analysis_doc) = node.document() else {
+            return Err(anyhow!(
+                "document analysis data not available for {}",
+                document_uri
+            ));
+        };
+
+        let offset = position_to_offset(&lines, position, encoding)?;
+        let Some(token) = find_identifier_token_at_offset(&root, offset) else {
+            return Err(anyhow!("no identifier found at position"));
+        };
+
+        let ident_text = token.text();
+        let parent_node = token
+            .parent()
+            .ok_or_else(|| anyhow!("identifier token has no parent"))?;
+
+        // Context based resolution
```

Comment:
We either need to refine or remove this comment, as it doesn't provide much insight.